### ISSUE TITLE
pandas-stubs works with 1.1.326, not with 1.1.329

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#6044]
Using pyright 1.1.326, our tests for pandas-stubs works with version 1.1.326.
With versions 1.1.327 and 1.1.328, there were failures, but those were fixed in 1.1.329.
But in 1.1.329, a new error appeared in `tests/test_indexes.py` .  Note - we are using numpy 1.26.0

This code in our tests:
```python
        assert_type(
            pd.IntervalIndex.from_breaks(
                np.array([1.0, 2.0, 3.0, 4.0], dtype=np.float64)
            ),
            "pd.IntervalIndex[pd.Interval[float]]",
        ),
```
reports an error with 1.1.329:  `error: "assert_type" mismatch: expected "IntervalIndex[Interval[float]]" but received "IntervalIndex[Interval[int]]" (reportGeneralTypeIssues)`  but not with earlier versions.

But if I do:
```python
        foo = np.array([1.0, 2.0, 3.0, 4.0], dtype=np.float64)
        assert_type(
            foo,
            "pd.IntervalIndex[pd.Interval[float]]",
        ),
```
then there is no error

I thought you were doing regression testing with pandas-stubs as part of CI now??

**Comment by erictraut** [comment]
Thanks for reporting.

When considering the impact of a bug fix or an enhancement, I run mypy_primer on several dozen libraries including pandas-stubs. You shouldn't assume that this will never result in changes, and some of these changes could result in type checking changes for some libraries.

The problem is likely related to an improvement I made in 1.1.329 relating to type checks for `float` and `int`. This was prompted by [this bug](https://github.com/microsoft/pyright/issues/6008).

I'll investigate and get back to you.

**Comment by Dr-Irv** [comment]
> When considering the impact of a bug fix or an enhancement, I run mypy_primer on several dozen libraries including pandas-stubs. You shouldn't assume that this will never result in changes, and some of these changes could result in type checking changes for some libraries.

Sure, but I guess if you do see a change in pyright that will affect `pandas-stubs`, if you create an issue, we can then figure out whether it is something we need to fix or not.

**Comment by erictraut** [comment]
My initial guess was wrong. This is caused by a change I made to an overload resolution heuristic as part of this [this change](https://github.com/microsoft/pyright/commit/afbe9aa1f4867d821b295d928a3f23448e3fc467).

In an effort to debug the change process... Here's the [mypy_primer run](https://github.com/microsoft/pyright/pull/6014) that shows the impact of this change to `pandas_stubs`. I remember looking at these results and convincing myself that all of the changes were innocuous (a difference in the wording of the error messages or `assert_type` failures that were failing previous to the change). But in that long list there was one important result that I missed.

The output of mypy_primer makes it easy to miss important results, especially when they're embedded in a sea of unimportant results. I can't think of any specific ways to improve it, so I guess that means I'll need to just get better (or spend more time) scrutinizing the results.

**Comment by erictraut** [comment]
Here's a thought... @Dr-Irv, would you be willing to take a look at how `mypy_primer` runs mypy and pyright on `pandas-stubs`? If we could make those runs cleaner (i.e. closer to the way you run the tests in your CI pipeline), it would help reduce the noise and make it more obvious when a legitimate regression occurs.

Here's where mypy_primer [configures its test for pandas-stubs](https://github.com/hauntsaninja/mypy_primer/blob/7092bbe9be7f653e152d8b60c15867fb360a1373/mypy_primer/projects.py#L813). Note that it doesn't currently install any dependencies that the pandas-stubs depend upon. For example, it doesn't install numpy. That significantly changes the results from what you normally see when you run the tests in the pandas-stubs repo.

@hauntsaninja, as maintainer of `mypy_primer`, do you have any thoughts or advice here?

**Comment by Dr-Irv** [comment]
Our CI runs it as `mypy pandas-stubs tests` and `pyright pandas-stubs tests` .  In the config you pointed to, I don't know why it looks like this:
```
            mypy_cmd="{mypy} pandas-stubs tests",
            pyright_cmd="{pyright}",
```
Some responses to your comments:

1. I would have thought it would be `pyright_cmd="{pyright} pandas-stubs tests"` .
2. I think it's important to install the dependencies, especially `numpy` because we are dependent on types in `numpy` (and some other python libraries as well) that we know will be installed when `pandas` is installed.  
3. I'm not sure how the `mypy_primer` works in terms of comparing output, but in the output you sent me, there are lines that were removed and lines that were added.  It's the ones that are added that you should be looking at.  IMHO, if you had run `pyright` using version 1.1.326 on pandas-stubs (with the right dependencies installed and the same pyright settings), you would have no errors, so that a comparison with an "in development" version would compare the results to the last known working version (nothing reported), and if new things were picked up, they would show up in green and are worth investigating (either by you or by us).
4. In our `pyproject.toml` file, we have a list of dependencies, but we also have the list of `pyright` settings that we use, and, ideally, you'd be using those settings when using `mypy_primer` with `pyright`

I should note that in our CI, when we do a merge, we run the latest "nightly" version of `mypy` so we are aware of changes that are coming in `mypy` that might affect us in the future.  As an example, the current in development version of mypy, it reports a new error message, which would be consistent with other messages that are reported with mypy 1.5.1, and we need to add a `#type: ignore` comment.  But if I put in that comment for the nightly version, then the 1.5.1 version complains.  So whenever the next release of mypy occurs, I'll then do a fix that includes that comment.

So I'm not sure if it's possible to incorporate such a test for a "nightly" version of pyright, but we'd be willing to do so.  It would only get run when a pull request is merged in, but it might catch new things as well for pyright like it does for mypy.

**Comment by erictraut** [comment]
> I'm not sure how the mypy_primer works in terms of comparing output

It simply runs the type checker twice — once prior to the change and once after the change, then it diffs the resulting outputs. In the results you looked at, the "before" was from prior to my commit and the "after" was with the commit applied.

> In our `pyproject.toml` file, we have a list of dependencies, but we also have the list of pyright settings that we use, and, ideally, you'd be using those settings when using mypy_primer with pyright

Pyright will pick up the settings from your `pyproject.toml`, so that's not a problem. But pyright doesn't install dependencies into your Python environment. It assumes that you have done this prior to invoking pyright. I don't think mypy_primer currently installs dependencies, at least not by default. Let's see if it's possible and feasible to add that.

> a test for a "nightly" version of pyright

We don't have the concept of a "nightly" build of pyright, but it would be possible to check out the latest source of pyright and build them as part of your CI script. This would add some overhead to your CI scripts — probably five minutes to fetch all the dependencies and build the pyright bundle, so I'm not sure if that's a big issue for you. I suppose it could be done as a separate action that's executed in parallel.

Is that comparable to what you do for mypy? To my knowledge, mypy doesn't publish a "nightly" version to pypi, so I'm guessing that you need to install it from source? Do you build it (to create a mypyc executable), or do you simply run the slow Python version directly from source?

If you'd like to build pyright from sources in a script, here's what you'd need to do:
1. Check out the latest "main" branch of `https://github.com/Microsoft/pyright`.
2. Make sure you're using node 16 (which is the version our build system assumes).
3. From the root directory, execute `npm i` to install all dependencies.
4. From the `packages/pyright` directory, execute `npm run build`.
5. To invoke the newly-built version of the pyright cli, execute `node pyright/packages/pyright/index.js <command-line options>`. Alternatively, you can install the newly-built version globally in your environment using `npm install . -g`, then invoke it using `pyright`.

**Comment by Dr-Irv** [comment]
> > I'm not sure how the mypy_primer works in terms of comparing output
> 
> It simply runs the type checker twice — once prior to the change and once after the change, then it diffs the resulting outputs. In the results you looked at, the "before" was from prior to my commit and the "after" was with the commit applied.

So the issue is the "before" version.  IMHO, the "before" should be the last one that ran successfully with no error output.  Or you just have to make sure that nothing changes from a released version to the first commit after a release.

> 
> Is that comparable to what you do for mypy? To my knowledge, mypy doesn't publish a "nightly" version to pypi, so I'm guessing that you need to install it from source? Do you build it (to create a mypyc executable), or do you simply run the slow Python version directly from source?

We just do `pip install --upgrade git+https://github.com/python/mypy.git` and run from that version.  It takes just over 3 minutes to run the whole job (installation plus type checking).  So I guess it is running the slow python version.  We only do this on approved pull requests so that contributors don't worry about failures.

> 
> If you'd like to build pyright from sources in a script, here's what you'd need to do:
> 
Created https://github.com/pandas-dev/pandas-stubs/issues/786

**Comment by erictraut** [comment]
> the "before" should be the last one that ran successfully with no error output

None of the `mypy_primer` runs for `python-stubs` produce "no error output" because they don't install dependencies. That's the fundamental issue here. Let's see if we can get that addressed.

**Comment by twoertwein** [comment]
Mypy primer has dependencies for [pandas](https://github.com/hauntsaninja/mypy_primer/blob/7092bbe9be7f653e152d8b60c15867fb360a1373/mypy_primer/projects.py#L98C43-L98C43)  but not for [pandas-stubs](https://github.com/hauntsaninja/mypy_primer/blob/7092bbe9be7f653e152d8b60c15867fb360a1373/mypy_primer/projects.py#L814). I would assume that installing numpy will significantly reduce the number of noisy errors that pollute the mypy primer output.

About having a "pyright nightly" run on pandas: @erictraut fixes bugs so quickly that I'm not sure whether it is worth the effort (that is a different for mypy, which has far more infrequent releases).

**Comment by erictraut** [comment]
This was a real regression. Apologies for not catching it sooner. It will be addressed in the next release.

**Comment by hauntsaninja** [comment]
Yeah, if the summary is `+ 1337 errors, 0 warnings, 0 informations` it's not representative of what you are running. I added more dependencies to the project definition of `pandas-stubs`, so hopefully things will be more representative going forward.

I had a mypy_primer mode where it would alert me if projects I expected clean type checking runs on were unclean (see the `expected_mypy_success` flag). But I haven't audited it in a while / have never done so for pyright.

Perhaps interestingly, I can only remember maybe two instances where a similar mismatch caused a miss in mypy CI (but also mypy_primer runs many more projects for mypy).

I agree Eric's / pyright's release cadence and turnaround time is quick enough that running pyright nightly might not get you too much.

mypy does not publish nightly versions to PyPI, but if you do want mypyc-compiled nightly wheels, you can find them at https://github.com/mypyc/mypy_mypyc-wheels

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
